python/mlx/nn/layers/positional_encoding.py

```diff
@@ -75,31 +90,43 @@ def __call__(self, x, offset: int = 0):
         shape = x.shape
         x = mx.reshape(x, (-1, shape[-2], shape[-1]))
         N = x.shape[1] + offset
-        costheta, sintheta = RoPE.create_cos_sin_theta(
-            N, self.dims, offset=offset, base=self.base, scale=self.scale, dtype=x.dtype
-        )
+        if N > self.max_position:
+            self._set_cos_sin_cache(
+                self.dims,
+                self.max_position,
+                seq_len=N,
+                base=self.base,
+                scale=self.scale,
+                dtype=x.dtype,
+            )
 
         rope = (
             self._compute_traditional_rope if self.traditional else self._compute_rope
         )
-        rx = rope(costheta, sintheta, x)
+
+        costheta, sintheta = self._cos_sin_theta_value
+        rx = rope(costheta[offset:N], sintheta[offset:N], x)
 
         return mx.reshape(rx, shape)
 
-    @staticmethod
-    def create_cos_sin_theta(
```

Comment:
rewrite the `create_cos_sin_theta` function because we only want to precompute the theta once.